### PR TITLE
Combined Encounter Models

### PR DESCRIPTION
**Pull request by leobcgov**
# Fixes or Implements [AB#9911](https://qslvic.visualstudio.com/304a1f8c-dace-4f85-adf3-bf563d5b3a39/_workitems/edit/9911)

* [X] Enhancement
* [ ] Bug
* [ ] Documentation

## Description
- Refactored Encounter Models for combining Claims.
image

If you are reviewing this PR, please adhere to the guidelines as [documented](https://github.com/bcgov/healthgateway/wiki/prguidance).

## Testing

* [X] Unit Tests Updated
* [ ] Functional Tests Updated
* [ ] Not Required

### Steps to Reproduce

If this is a bug, please provide details on how to reproduce the code.

### UI Changes

NO

### Browsers Tested

* [X] Chrome
* [ ] Safari
* [ ] Edge
* [ ] Firefox

Provide an explanation for any test(s) not completed.

## Notes/Additional Comments

Any additional notes or comments that may be relevant.  Include any specialized deployment steps here.

### CODE REVIEW

File: Apps/Encounter/src/Models/EncounterModel.cs [hunk 1]
```diff
@@ -89,10 +90,16 @@ public static EncounterModel FromODRClaimModel(Claim model)
         public static List<EncounterModel> FromODRClaimModelList(List<Claim> models)
         {
             List<EncounterModel> objects = new List<EncounterModel>();
-
+            Dictionary<string, EncounterModel> dictEncounterIds =
```
Comment (FuriousLlama) [inline Apps/Encounter/src/Models/EncounterModel.cs:93]: We are not using the value from the dictionary. Does it need to be a dictionary?

File: Apps/Common/src/Utils/HashGenerator.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+namespace HealthGateway.Common.Utils
+{
+    using System;
+    using System.Security.Cryptography;
+    using System.Text;
+
+    /// <summary>
+    /// Utilities for Hash.
+    /// </summary>
+    public static class HashGenerator
```
Comment (FuriousLlama) [inline Apps/Common/src/Utils/HashGenerator.cs:26]: not sure this class is needed. So far is only being used once and for a very specific purpose

File: Apps/Encounter/src/Models/EncounterModel.cs [hunk 3]
```diff
@@ -63,7 +64,7 @@ public static EncounterModel FromODRClaimModel(Claim model)
         {
             return new EncounterModel()
             {
-                Id = model.ClaimId.ToString(),
+                Id = HashGenerator.ComputeHashToGuid($"{model.ServiceDate:yyyyMMdd}{model.SpecialtyDesc}{model.PractitionerName}{model.LocationName}").ToString(),
```
Comment (FuriousLlama) [inline Apps/Encounter/src/Models/EncounterModel.cs:67]: I think we also need to qualify it by the clinic full address

File: Apps/Common/test/unit/Utils/HashGeneratorTest.cs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    {
+        /// <summary>
+        /// Should Compute Hash To Guid.
+        /// </summary>
+        [Fact]
+        public void ShouldComputeHashToGuid()
+        {
+            var aLongString = "17ADDA2E26EE6E8EA930A51577026E9E67FBDF46EC175C8DEFD3EAB08437AE0";
+            var generatedGuid = HashGenerator.ComputeHashToGuid(aLongString);
+            Assert.NotEqual(generatedGuid.ToString(), Guid.Empty.ToString());
```
Comment (FuriousLlama) [inline Apps/Common/test/unit/Utils/HashGeneratorTest.cs:39]: needs to test that it generates the same guid as well

File: Apps/Encounter/src/Models/EncounterModel.cs [hunk 5]
```diff
@@ -18,6 +18,7 @@ namespace HealthGateway.Encounter.Models
     using System;
     using System.Collections.Generic;
     using System.Text.Json.Serialization;
+    using HealthGateway.Common.Utils;
```
Comment (FuriousLlama) [inline Apps/Encounter/src/Models/EncounterModel.cs:21]: "using" order smell

File: Apps/Encounter/src/Models/EncounterModel.cs (hunk 1 above)
Comment (leobcgov) [reply Apps/Encounter/src/Models/EncounterModel.cs:93]: @FuriousLlama what is your suggestion?

File: Apps/Encounter/src/Models/EncounterModel.cs (hunk 3 above)
Comment (leobcgov) [reply Apps/Encounter/src/Models/EncounterModel.cs:67]: @ninosamson can you clarify this?

File: Apps/Encounter/src/Models/EncounterModel.cs (hunk 1 above)
Comment (FuriousLlama) [reply Apps/Encounter/src/Models/EncounterModel.cs:93]: HashSet, for example:
https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.hashset-1?view=net-5.0

Comment (FuriousLlama) [reply Apps/Encounter/src/Models/EncounterModel.cs:93]: also: https://stackoverflow.com/questions/1378819/using-a-hashtable-to-store-only-keys

File: Apps/Encounter/src/Models/EncounterModel.cs (hunk 3 above)
Comment (leobcgov) [reply Apps/Encounter/src/Models/EncounterModel.cs:67]: per @ninosamson 's confirmation, I will include the full address.

File: Apps/Encounter/src/Models/EncounterModel.cs (hunk 1 above)
Comment (leobcgov) [reply Apps/Encounter/src/Models/EncounterModel.cs:93]: good one

File: Apps/Encounter/src/Models/EncounterModel.cs [hunk 6]
```diff
@@ -61,9 +63,15 @@ public class EncounterModel
...
+#pragma warning disable CA5351 // Do Not Use Broken Cryptographic Algorithms
+#pragma warning disable SCS0006 // Weak hashing function
+            using var md5CryptoService = MD5.Create();
+#pragma warning restore SCS0006 // Weak hashing function
+#pragma warning restore CA5351 // Do Not Use Broken Cryptographic Algorithms
+
             return new EncounterModel()
             {
-                Id = model.ClaimId.ToString(),
+                Id = new Guid(md5CryptoService.ComputeHash(Encoding.Default.GetBytes($"{model.ServiceDate:yyyyMMdd}{model.SpecialtyDesc}{model.PractitionerName}{model.LocationName}{model.LocationAddress.Province}{model.LocationAddress.City}{model.LocationAddress.PostalCode}{model.LocationAddress.AddrLine1}{model.LocationAddress.AddrLine2}{model.LocationAddress.AddrLine3}{model.LocationAddress.AddrLine4}"))).ToString(),
```
Comment (FuriousLlama) [inline Apps/Encounter/src/Models/EncounterModel.cs:74]: nit, for readability I would use a string builder and add each of the qualifiers parameters

File: Apps/Encounter/src/Models/EncounterModel.cs [hunk 7]
```diff
@@ -89,10 +97,15 @@ public static EncounterModel FromODRClaimModel(Claim model)
         public static List<EncounterModel> FromODRClaimModelList(List<Claim> models)
         {
             List<EncounterModel> objects = new List<EncounterModel>();
-
+            var hsEncounterIds = new HashSet<string>();
```
Comment (FuriousLlama) [inline Apps/Encounter/src/Models/EncounterModel.cs:100]: nit: HashSet<string> encounterIds = new HashSet<string>();

File: Apps/Encounter/src/Models/EncounterModel.cs (hunk 6 above)
Comment (leobcgov) [reply Apps/Encounter/src/Models/EncounterModel.cs]: I prefer a 1 line code instead as you can configure to have wrap mode in VS, but it is fine :-) Updated

**Review by FuriousLlama — APPROVED** [review]

File: Apps/Encounter/src/Models/EncounterModel.cs (hunk 1 above)
Comment (sslaws) [reply Apps/Encounter/src/Models/EncounterModel.cs]: NIT:  We should be using tasks/9911 for the branch name now that we're using proper feature branches.

File: Apps/Encounter/src/Models/EncounterModel.cs [hunk 8]
```diff
@@ -61,9 +63,26 @@ public class EncounterModel
...
+#pragma warning disable SCS0006 // Weak hashing function
+            using var md5CryptoService = MD5.Create();
+#pragma warning restore SCS0006 // Weak hashing function
+#pragma warning restore CA5351 // Do Not Use Broken Cryptographic Algorithms
+            StringBuilder sourceId = new StringBuilder();
+            sourceId.Append($"{model.ServiceDate:yyyyMMdd}");
+            sourceId.Append($"{model.SpecialtyDesc}");
+            sourceId.Append($"{model.PractitionerName}");
+            sourceId.Append($"{model.LocationName}");
+            sourceId.Append($"{model.LocationAddress.Province}");
```
Comment (sslaws) [inline Apps/Encounter/src/Models/EncounterModel.cs:76]: Seems like overkill to continue after the location name.

**Review by sslaws — APPROVED** [review]

**Pull request merged**